File: backend/src/main/java/com/workout/app/service/DailyTaskService.java
package com.workout.app.service;

import com.workout.app.entity.DailyTask;
import com.workout.app.repository.DailyTaskRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.List;

@Service
public class DailyTaskService {

    @Autowired
    private DailyTaskRepository dailyTaskRepository;

    public List<DailyTask> getUserTasks(Long userId) {
        return dailyTaskRepository.findByUserId(userId);
    }

    public List<DailyTask> getIncompleteTasks(Long userId) {
        return dailyTaskRepository.findByUserIdAndCompleted(userId, false);
    }

    public DailyTask createTask(Long userId, String title) {
        DailyTask task = new DailyTask();
        task.setUserId(userId);
        task.setTitle(title);
        task.setCompleted(false);
        task.setCreatedAt(LocalDateTime.now());
        return dailyTaskRepository.save(task);
    }

    public DailyTask toggleTask(Long taskId) {
        DailyTask task = dailyTaskRepository.findById(taskId)
                .orElseThrow(() -> new RuntimeException("Task not found"));
        task.setCompleted(!task.isCompleted());
        return dailyTaskRepository.save(task);
    }

    public void deleteTask(Long taskId) {
        dailyTaskRepository.deleteById(taskId);
    }

    public void resetAllTasks(Long userId) {
        List<DailyTask> tasks = dailyTaskRepository.findByUserId(userId);
        LocalDateTime now = LocalDateTime.now();

        for (DailyTask task : tasks) {
            task.setCompleted(false);
            task.setLastResetAt(now);
        }

        dailyTaskRepository.saveAll(tasks);
    }

    public void checkAndResetTasksIfNeeded(Long userId) {
        List<DailyTask> tasks = dailyTaskRepository.findByUserId(userId);

        if (tasks.isEmpty()) {
            return;
        }

        LocalDateTime now = LocalDateTime.now();
        LocalDate today = now.toLocalDate();

        // Check if any task was reset today
        boolean needsReset = tasks.stream()
                .noneMatch(task -> {
                    LocalDateTime lastReset = task.getLastResetAt();
                    return lastReset != null && lastReset.toLocalDate().equals(today);
                });

        if (needsReset) {
            resetAllTasks(userId);
        }
    }
}

File: backend-node/src/services/DailyTaskService.ts
import { DailyTask } from '@prisma/client';
import prisma from '../lib/database';

export class DailyTaskService {
  async getUserTasks(userId: bigint): Promise<DailyTask[]> {
    return prisma.dailyTask.findMany({
      where: { userId },
      orderBy: { createdAt: 'desc' }
    });
  }

  async getIncompleteTasks(userId: bigint): Promise<DailyTask[]> {
    return prisma.dailyTask.findMany({
      where: { 
        userId,
        completed: false
      },
      orderBy: { createdAt: 'desc' }
    });
  }

  async createTask(userId: bigint, title: string): Promise<DailyTask> {
    return prisma.dailyTask.create({
      data: {
        userId,
        title,
        completed: false,
        createdAt: new Date()
      }
    });
  }

  async toggleTask(taskId: bigint): Promise<DailyTask> {
    const task = await prisma.dailyTask.findUnique({
      where: { id: taskId }
    });

    if (!task) {
      throw new Error('Task not found');
    }

    return prisma.dailyTask.update({
      where: { id: taskId },
      data: { completed: !task.completed }
    });
  }

  async deleteTask(taskId: bigint): Promise<void> {
    await prisma.dailyTask.delete({
      where: { id: taskId }
    });
  }

  async resetAllTasks(userId: bigint): Promise<void> {
    const now = new Date();
    
    await prisma.dailyTask.updateMany({
      where: { userId },
      data: {
        completed: false,
        lastResetAt: now
      }
    });
  }

  async checkAndResetTasksIfNeeded(userId: bigint): Promise<void> {
    const tasks = await this.getUserTasks(userId);

    if (tasks.length === 0) {
      return;
    }

    const now = new Date();
    const today = now.toISOString().split('T')[0]; // Get YYYY-MM-DD

    // Check if any task was reset today
    const needsReset = !tasks.some(task => {
      if (!task.lastResetAt) return false;
      const lastResetDate = task.lastResetAt.toISOString().split('T')[0];
      return lastResetDate === today;
    });

    if (needsReset) {
      await this.resetAllTasks(userId);
    }
  }
}